category/traversable/derive.lean

```diff
@@ -0,0 +1,208 @@
+/-
+Copyright (c) 2018 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Simon Hudon
+
+Automation to construct `traversable` instances
+-/
+
+import category.traversable.basic category.traversable.instances category.basic
+import tactic.basic tactic.cache
+
+namespace tactic.interactive
+
+open tactic list monad functor
+
+meta def nested_map (f v : expr) : expr → tactic expr
+| t :=
+do t ← instantiate_mvars t,
+   mcond (succeeds $ is_def_eq t v)
+      (pure f)
+      (if ¬ v.occurs (t.app_fn)
+          then do
+            cl ← mk_app ``functor [t.app_fn],
+            _inst ← mk_instance cl,
+            f' ← nested_map t.app_arg,
+            mk_mapp ``functor.map [t.app_fn,_inst,none,none,f']
+          else fail format!"type {t} is not a functor with respect to variable {v}")
+
+meta def map_field (n : name) (cl f v e : expr) : tactic expr :=
+do t ← infer_type e >>= whnf,
+   if t.get_app_fn.const_name = n
+   then fail "recursive types not supported"
+   else if v.occurs t
+   then do f' ← nested_map f v t,
+           pure $ f' e
+   else
+         (is_def_eq t.app_fn cl >> to_expr ``(comp.mk %%e))
+     <|> pure e
+
+meta def nested_traverse (f v : expr) : expr → tactic expr
+| t :=
+do t ← instantiate_mvars t,
+   mcond (succeeds $ is_def_eq t v)
+      (pure f)
+      (if ¬ v.occurs (t.app_fn)
+          then do
+            cl ← mk_app ``traversable [t.app_fn],
+            _inst ← mk_instance cl,
+            f' ← nested_traverse t.app_arg,
+            mk_mapp ``traversable.traverse [t.app_fn,_inst,none,none,none,none,f']
+          else fail format!"type {t} is not traversable with respect to variable {v}")
+
+meta def traverse_field (n : name) (inst cl f v e : expr) : tactic expr :=
+do t ← infer_type e >>= whnf,
+   if t.get_app_fn.const_name = n
+   then fail "recursive types not supported"
+   else if v.occurs t
+   then do f' ← nested_traverse f v t,
+           pure $ f' e
+   else
+         (is_def_eq t.app_fn cl >> to_expr ``(comp.mk %%e))
+     <|> to_expr ``(@pure _ (%%inst).to_has_pure _ (ulift.up %%e))
+
+meta def seq_apply_constructor : list expr → expr → tactic expr
+| (x :: xs) e := to_expr ``(%%e <*> %%x) >>= seq_apply_constructor xs
+| [] e := return e
+
+meta def fill_implicit_arg' : expr → expr → tactic expr
+| f (expr.pi n bi d b) :=
+if b.has_var then
+do v ← mk_meta_var d,
+   fill_implicit_arg' (expr.app f v) (b.instantiate_var v)
+else return f
+| e _ := return e
+
+meta def fill_implicit_arg (n : name) : tactic expr :=
+do c ← mk_const n,
+   t ← infer_type c,
+   fill_implicit_arg' c t
+
+meta def mk_down (e : expr) : tactic expr :=
+to_expr ``(ulift.down %%e) <|> pure e
```

Comment:
I think `mk_app ``ulift.down [e]` should perform the same action without the full elaborator getting in on the action